File: python/deepstrike/providers/anthropic_adapter.py
"""Anthropic Messages protocol lifecycle adapter."""
from __future__ import annotations

import json
from dataclasses import dataclass, field
from typing import Any

from deepstrike._kernel import Message, ToolCall, ToolSchema
from deepstrike.providers.base import normalize_tool_call, parse_tool_arguments, to_anthropic_messages
from deepstrike.providers.protocol_adapter import AdapterOutput, ProtocolResponseError
from deepstrike.providers.stop_reason import canonicalize_stop_reason
from deepstrike.providers.stream import TextDelta, ThinkingDelta, ToolCallEvent, UsageEvent
from deepstrike.providers.usage import ProviderUsage
from deepstrike.types.content import CanonicalAdapterInput


def _get(raw: Any, key: str) -> Any:
    return raw.get(key) if isinstance(raw, dict) else getattr(raw, key, None)


def _number(raw: Any, field: str) -> int | None:
    value = _get(raw, field)
    if value is None:
        return None
    if isinstance(value, bool) or not isinstance(value, (int, float)) or value < 0:
        raise ProtocolResponseError("anthropic-messages", f"usage.{field} is invalid")
    return int(value)


@dataclass(frozen=True)
class AnthropicRequestPlan:
    transport: str
    params: dict[str, Any]
    cache_slots: dict[str, bool]


@dataclass
class AnthropicStreamState:
    native_blocks: dict[int, dict] = field(default_factory=dict)
    tool_blocks: dict[int, dict] = field(default_factory=dict)
    final_tool_calls: list[ToolCall] = field(default_factory=list)
    final_text: str = ""
    uncached_input: int = 0
    cache_read: int = 0
    cache_creation: int = 0
    output_tokens: int = 0
    cache_slots: dict[str, bool] = field(default_factory=dict)


class AnthropicMessagesAdapter:
    protocol = "anthropic-messages"

    def __init__(self, model: str = "claude-sonnet-4-6") -> None:
        self._model = model

    def build_request(
        self,
        input: CanonicalAdapterInput,
        *,
        messages: list[dict] | None = None,
        system: Any = None,
        tools: list[dict] | None = None,
    ) -> AnthropicRequestPlan:
        context = input.context
        if messages is None:
            messages = to_anthropic_messages(context.turns, resolved=input.resolved)
        if not messages:
            messages = [{"role": "user", "content": "Proceed."}]
        if system is None:
            stable = getattr(context, "system_stable", "") or ""
            knowledge = getattr(context, "system_knowledge", "") or ""
            if stable or knowledge:
                cache_control = {"type": "ephemeral", **({"ttl": "1h"} if input.extensions.get("cacheTtl") == "1h" else {})}
                system = [
                    {"type": "text", "text": text, "cache_control": cache_control}
                    for text in (stable, knowledge) if text
                ]
            else:
                system = context.system_text or None
        if tools is None and input.tools:
            tools = [{"name": tool.name, "description": tool.description, "input_schema": json.loads(tool.parameters)} for tool in input.tools]
        ext = {key: value for key, value in input.extensions.items() if key not in {"model", "messages", "system", "tools", "stream", "max_tokens", "cacheBreakpointStrategy", "cacheTtl"}}
        betas = input.extensions.get("betas")
        if isinstance(betas, list) and betas:
            ext["betas"] = betas
        params = {
            **ext,
            "model": input.resolved.model_id if input.resolved is not None else self._model,
            "max_tokens": input.extensions.get("max_tokens", 8096),
            **({"system": system} if system else {}),
            "messages": messages,
            **({"tools": tools} if tools else {}),
        }
        return AnthropicRequestPlan(
            transport="beta" if betas else "stable",
            params=params,
            cache_slots={
                "system": isinstance(system, list) and any(isinstance(block, dict) and block.get("cache_control") for block in system),
                "tools": bool(tools) and any(isinstance(tool, dict) and tool.get("cache_control") for tool in tools or []),
                "messages": any(isinstance(message.get("content"), list) and any(block.get("cache_control") for block in message["content"] if isinstance(block, dict)) for message in messages),
            },
        )

    def decode_complete(self, raw: Any, input: CanonicalAdapterInput) -> tuple[Message, dict | None]:
        content = ""
        calls: list[ToolCall] = []
        native_blocks: list[dict] = []
        for block in _get(raw, "content") or []:
            kind = _get(block, "type")
            if kind == "text":
                text = _get(block, "text") or ""
                content += text
                native_blocks.append({"type": "text", "text": text})
            elif kind == "tool_use":
                call = normalize_tool_call(_get(block, "id"), _get(block, "name"), _get(block, "input"))
                if call:
                    calls.append(call)
                native_blocks.append({"type": "tool_use", "id": _get(block, "id"), "name": _get(block, "name"), "input": _get(block, "input") or {}})
            elif kind == "thinking":
                native_blocks.append({"type": "thinking", "thinking": _get(block, "thinking"), "signature": _get(block, "signature")})
        usage = self.normalize_usage(_get(raw, "usage"))
        token_count = usage.input_tokens + usage.output_tokens if usage else None
        message = Message(role="assistant", content=content, token_count=token_count, tool_calls=calls or None)
        return message, ({"native_blocks": native_blocks} if native_blocks else None)

    def create_stream_state(self, input: CanonicalAdapterInput, cache_slots: dict[str, bool] | None = None) -> AnthropicStreamState:
        return AnthropicStreamState(cache_slots=dict(cache_slots or {}))

    def push_stream_chunk(self, chunk: Any, state: AnthropicStreamState) -> AdapterOutput:
        events = []
        kind = _get(chunk, "type")
        if kind in ("message_start", "message_delta"):
            usage = _get(chunk, "usage") or _get(_get(chunk, "message"), "usage")
            if usage is not None:
                state.uncached_input = max(state.uncached_input, _number(usage, "input_tokens") or 0)
                state.cache_read = max(state.cache_read, _number(usage, "cache_read_input_tokens") or 0)
                state.cache_creation = max(state.cache_creation, _number(usage, "cache_creation_input_tokens") or 0)
                state.output_tokens = max(state.output_tokens, _number(usage, "output_tokens") or 0)
                raw_stop = _get(_get(chunk, "delta"), "stop_reason")
                total_input = state.uncached_input + state.cache_read + state.cache_creation
                provider_usage = ProviderUsage(input_tokens=total_input, output_tokens=state.output_tokens, cache_read_input_tokens=state.cache_read, cache_creation_input_tokens=state.cache_creation)
                contributors = [key for key in ("system", "tools", "messages") if state.cache_slots.get(key)]
                by_slot = None
                if state.cache_read and contributors:
                    share, remainder = divmod(state.cache_read, len(contributors))
                    by_slot = {key: share + (remainder if index == 0 else 0) for index, key in enumerate(contributors)}
                events.append(UsageEvent(total_tokens=total_input + state.output_tokens, input_tokens=total_input, output_tokens=state.output_tokens, cache_read_input_tokens=state.cache_read, cache_creation_input_tokens=state.cache_creation, cache_read_input_tokens_by_slot=by_slot, stop_reason=canonicalize_stop_reason(raw_stop), raw_stop_reason=raw_stop, provider_usage=provider_usage))
        elif kind == "content_block_start":
            idx = int(_get(chunk, "index"))
            block = _get(chunk, "content_block")
            block_type = _get(block, "type")
            state.native_blocks[idx] = {"type": block_type}
            if block_type == "thinking":
                state.native_blocks[idx].update({"thinking": _get(block, "thinking") or "", "signature": _get(block, "signature") or ""})
            elif block_type == "text":
                state.native_blocks[idx]["text"] = _get(block, "text") or ""
            elif block_type == "tool_use":
                state.native_blocks[idx].update({"id": _get(block, "id"), "name": _get(block, "name"), "input": _get(block, "input") or {}})
                state.tool_blocks[idx] = {"id": _get(block, "id"), "name": _get(block, "name"), "args": ""}
        elif kind == "content_block_delta":
            idx = int(_get(chunk, "index"))
            delta = _get(chunk, "delta")
            delta_type = _get(delta, "type")
            if delta_type == "text_delta":
                text = _get(delta, "text") or ""
                state.final_text += text
                state.native_blocks.setdefault(idx, {"type": "text"})["text"] = state.native_blocks.get(idx, {}).get("text", "") + text
                events.append(TextDelta(delta=text))
            elif delta_type == "thinking_delta":
                text = _get(delta, "thinking") or ""
                state.native_blocks.setdefault(idx, {"type": "thinking"})["thinking"] = state.native_blocks.get(idx, {}).get("thinking", "") + text
                events.append(ThinkingDelta(delta=text))
            elif delta_type == "signature_delta":
                state.native_blocks.setdefault(idx, {"type": "thinking"})["signature"] = _get(delta, "signature") or ""
            elif delta_type == "input_json_delta" and idx in state.tool_blocks:
                state.tool_blocks[idx]["args"] += _get(delta, "partial_json") or ""
        elif kind == "content_block_stop":
            idx = int(_get(chunk, "index"))
            if idx in state.tool_blocks:
                block = state.tool_blocks.pop(idx)
                try:
                    args = json.loads(block["args"] or "{}")
                except json.JSONDecodeError:
                    args = {}
                state.native_blocks[idx]["input"] = args
                call = normalize_tool_call(block["id"], block["name"], args)
                if call:
                    state.final_tool_calls.append(call)
                    events.append(ToolCallEvent(id=call.id, name=call.name, arguments=args))
        return AdapterOutput(events=events)

    def finish_stream(self, state: AnthropicStreamState, final: Any = None) -> AdapterOutput:
        blocks = [state.native_blocks[index] for index in sorted(state.native_blocks)]
        return AdapterOutput(events=[], replay={"native_blocks": blocks} if blocks else None)

    def normalize_usage(self, raw: Any) -> ProviderUsage | None:
        if raw is None:
            return None
        if not isinstance(raw, dict) and not hasattr(raw, "input_tokens"):
            raise ProtocolResponseError("anthropic-messages", "usage must be an object")
        input_tokens = _number(raw, "input_tokens")
        cache_read = _number(raw, "cache_read_input_tokens")
        cache_creation = _number(raw, "cache_creation_input_tokens")
        output = _number(raw, "output_tokens")
        if input_tokens is None and cache_read is None and cache_creation is None and output is None:
            return None
        return ProviderUsage(input_tokens=(input_tokens or 0) + (cache_read or 0) + (cache_creation or 0), output_tokens=output or 0, cache_read_input_tokens=cache_read or 0, cache_creation_input_tokens=cache_creation or 0)

File: node/src/providers/anthropic-adapter.ts
import type {
  CacheBreakpointStrategy,
  ModelMessage,
  ProviderReplay,
  ProviderUsage,
  StreamEvent,
  ThinkingDelta,
  ToolCall,
  ToolCallEvent,
  UsageEvent,
} from "../types.js"
import type {
  CanonicalAdapterInput,
  CanonicalMessage,
  CanonicalToolResult,
} from "./content-normalization.js"
import { projectToolOutputToText } from "./content-normalization.js"
import { normalizeToolCall } from "./base.js"
import {
  type AdapterDecodeInput,
  type AdapterOutput,
  type AdapterStreamInput,
  type CanonicalStopReason,
  type ProtocolAdapter,
  ProtocolResponseError,
} from "./protocol-adapter.js"
import { ANTHROPIC_PROTOCOL_CAPABILITIES } from "./protocol-capabilities.js"
import { endpointProfiles } from "./endpoints.js"

export const ANTHROPIC_TEXTUAL_TOOL_CALL_START_MARKER = "<｜｜DSML｜｜tool_calls>"
const TEXTUAL_TOOL_CALL_CAPTURE_LIMIT = 64 * 1024

type TextualToolCallPolicy = "off" | "reject"

interface TextualToolCallGuardState {
  policy: TextualToolCallPolicy
  mode: "passthrough" | "candidate"
  tail: string
  capturedLength: number
  sawNativeToolCall: boolean
}

export interface AnthropicRequestPlan {
  transport: "stable" | "beta"
  params: Record<string, unknown>
  cacheSlots: { system: boolean; tools: boolean; messages: boolean }
}

export type AnthropicStreamChunk = Record<string, any>

export interface AnthropicStreamState {
  readonly input: CanonicalAdapterInput
  readonly cacheSlots: { system: boolean; tools: boolean; messages: boolean }
  readonly toolBlocks: Record<number, { id: string; name: string; argsBuffer: string }>
  readonly nativeBlocks: Record<number, Record<string, unknown>>
  readonly finalToolCalls: ToolCall[]
  readonly textualToolCallGuard: TextualToolCallGuardState
  finalText: string
  uncachedInput: number
  cacheReadTokens: number
  cacheCreationTokens: number
  cacheTelemetryMeasured: boolean
  outputTokens: number
}

export interface AnthropicStreamChunkRequest {
  chunk: AnthropicStreamChunk
  state: AnthropicStreamState
}

export interface AnthropicStreamFinishRequest {
  state: AnthropicStreamState
}

const CACHE_BREAKPOINT_STRATEGIES = new Set<CacheBreakpointStrategy>([
  "default", "tools-only", "system-only", "frozen-prefix", "none",
])

function cacheStrategy(extensions: Readonly<Record<string, unknown>>): CacheBreakpointStrategy {
  const raw = extensions.cacheBreakpointStrategy
  return typeof raw === "string" && CACHE_BREAKPOINT_STRATEGIES.has(raw as CacheBreakpointStrategy)
    ? raw as CacheBreakpointStrategy
    : "default"
}

function extensionsForWire(
  extensions: Readonly<Record<string, unknown>>,
): Record<string, unknown> {
  const blocked = new Set([
    "model", "messages", "system", "tools", "max_tokens", "stream",
    "__deepstrikeThinkingEnabled", "degradeMissingReasoningReplay", "textualToolCallPolicy",
  ])
  return Object.fromEntries(Object.entries(extensions).filter(([key]) => !blocked.has(key)))
}

function textualToolCallPolicy(input: CanonicalAdapterInput): TextualToolCallPolicy {
  if (input.tools.length === 0) return "off"
  const explicit = input.extensions.textualToolCallPolicy
  if (explicit === "off" || explicit === "reject") return explicit
  const official = endpointProfiles["anthropic.messages"]
  const endpoint = input.resolved.endpoint
  const isOfficial = input.resolved.identity.endpointId === official.id
    && (!endpoint || endpoint.baseURL === official.baseURL)
  return isOfficial ? "off" : "reject"
}

function textualToolCallError(): ProtocolResponseError {
  return new ProtocolResponseError(
    "anthropic-messages",
    "Provider emitted a tool call as text instead of a native tool block",
    { providerCode: "textual_tool_call", retryable: true },
  )
}

function hasTextualToolCall(text: string, input: CanonicalAdapterInput): boolean {
  return textualToolCallPolicy(input) === "reject"
    && text.includes(ANTHROPIC_TEXTUAL_TOOL_CALL_START_MARKER)
}

function markerTailLength(text: string): number {
  const max = Math.min(text.length, ANTHROPIC_TEXTUAL_TOOL_CALL_START_MARKER.length - 1)
  for (let length = max; length > 0; length -= 1) {
    if (text.endsWith(ANTHROPIC_TEXTUAL_TOOL_CALL_START_MARKER.slice(0, length))) return length
  }
  return 0
}

function utf8Length(text: string): number {
  return Buffer.byteLength(text, "utf8")
}

function guardTextDelta(text: string, guard: TextualToolCallGuardState): string {
  if (guard.policy === "off") return text
  if (guard.mode === "candidate") {
    guard.capturedLength += utf8Length(text)
    if (guard.capturedLength > TEXTUAL_TOOL_CALL_CAPTURE_LIMIT) throw textualToolCallError()
    return ""
  }
  const combined = guard.tail + text
  const markerIndex = combined.indexOf(ANTHROPIC_TEXTUAL_TOOL_CALL_START_MARKER)
  if (markerIndex >= 0) {
    guard.mode = "candidate"
    guard.tail = ""
    guard.capturedLength = utf8Length(combined.slice(markerIndex))
    if (guard.capturedLength > TEXTUAL_TOOL_CALL_CAPTURE_LIMIT) throw textualToolCallError()
    return combined.slice(0, markerIndex)
  }
  const tailLength = markerTailLength(combined)
  guard.tail = tailLength > 0 ? combined.slice(-tailLength) : ""
  return tailLength > 0 ? combined.slice(0, -tailLength) : combined
}

function systemBlocks(
  context: CanonicalAdapterInput["context"],
  strategy: CacheBreakpointStrategy,
): string | Array<Record<string, unknown>> | undefined {
  if (!context.systemStable && !context.systemKnowledge) {
    return context.systemText || undefined
  }
  const cache = strategy === "default" || strategy === "system-only"
  const blocks: Array<Record<string, unknown>> = []
  if (context.systemStable) {
    blocks.push({
      type: "text",
      text: context.systemStable,
      ...(cache ? { cache_control: { type: "ephemeral" } } : {}),
    })
  }
  if (context.systemKnowledge) {
    blocks.push({
      type: "text",
      text: context.systemKnowledge,
      ...(cache ? { cache_control: { type: "ephemeral" } } : {}),
    })
  }
  return blocks.length ? blocks : undefined
}

function toolResult(result: CanonicalToolResult): Record<string, unknown> {
  const blocks = result.blocks
  const native = blocks.map(block => {
    if (block.type === "text") return { type: "text", text: block.text }
    if (block.type === "image") {
      if (block.source.kind === "base64") {
        return {
          type: "image",
          source: {
            type: "base64",
            media_type: block.mediaType ?? "image/png",
            data: block.source.data,
          },
        }
      }
      if (block.source.kind === "url") {
        return { type: "image", source: { type: "url", url: block.source.url } }
      }
    }
    return { type: "text", text: `[${block.type}]` }
  })
  return {
    type: "tool_result",
    tool_use_id: result.callId,
    content: result.contentForm !== "blocks" && blocks.length === 1 && blocks[0].type === "text"
      ? blocks[0].text
      : native,
    is_error: result.isError,
  }
}

function canonicalMessage(message: CanonicalMessage): Record<string, unknown> | undefined {
  if (
    message.role === "assistant"
    && message.toolCalls?.length
    && message.providerReplay?.native_blocks?.length
  ) {
    return { role: "assistant", content: ensureAssistantToolText(message.providerReplay.native_blocks) }
  }
  const content: Array<Record<string, unknown>> = []
  for (const item of message.blocks) {
    if (item.type === "tool_result") {
      content.push(toolResult(item))
    } else if (item.type === "text") {
      if (item.text) content.push({ type: "text", text: item.text })
    } else if (item.type === "image") {
      if (item.source.kind === "base64") {
        content.push({
          type: "image",
          source: {
            type: "base64",
            media_type: item.mediaType ?? "image/png",
            data: item.source.data,
          },
        })
      } else if (item.source.kind === "url") {
        content.push({ type: "image", source: { type: "url", url: item.source.url } })
      }
    }
  }
  if (message.role === "assistant" && message.toolCalls?.length) {
    if (!content.length) content.push({ type: "text", text: "Tool call requested." })
    for (const call of message.toolCalls) {
      let args: Record<string, unknown> = {}
      try { args = JSON.parse(call.arguments) as Record<string, unknown> } catch { args = {} }
      content.push({ type: "tool_use", id: call.id, name: call.name, input: args })
    }
  }
  if (message.role === "tool" && !content.some(block => block.type === "tool_result")) {
    return undefined
  }
  return {
    role: message.role === "tool" ? "user" : message.role,
    content: message.contentForm !== "blocks" && content.length === 1 && content[0].type === "text"
      ? content[0].text
      : content,
  }
}

function ensureAssistantToolText(
  blocks: Array<Record<string, unknown>>,
): Array<Record<string, unknown>> {
  if (!blocks.some(block => block.type === "tool_use")) return blocks
  if (blocks.some(block => block.type === "text" && String(block.text ?? "").trim())) return blocks
  if (blocks.some(block => block.type === "thinking")) return blocks
  return [{ type: "text", text: "Tool call requested." }, ...blocks]
}

function markLastBlockCacheable(message: Record<string, unknown>): void {
  const cache_control = { type: "ephemeral" }
  if (typeof message.content === "string") {
    if (message.content) {
      message.content = [{ type: "text", text: message.content, cache_control }]
    }
    return
  }
  if (Array.isArray(message.content) && message.content.length) {
    const last = message.content.at(-1) as Record<string, unknown>
    last.cache_control = cache_control
  }
}

function applyMessageCacheControl(
  messages: Array<Record<string, unknown>>,
  frozenPrefixLen: number | undefined,
  strategy: CacheBreakpointStrategy,
): void {
  if (!messages.length || ["tools-only", "system-only", "none"].includes(strategy)) return
  const targets = new Set([messages.length - 1])
  if (typeof frozenPrefixLen === "number" && frozenPrefixLen >= 1 && frozenPrefixLen < messages.length) {
    targets.add(frozenPrefixLen - 1)
  } else if (strategy === "default") {
    for (let index = messages.length - 2; index >= 0 && targets.size < 2; index--) {
      if (messages[index].role === "user") targets.add(index)
    }
  }
  for (const index of targets) markLastBlockCacheable(messages[index])
}

function requestTools(
  input: CanonicalAdapterInput,
  system: string | Array<Record<string, unknown>> | undefined,
  strategy: CacheBreakpointStrategy,
): Array<Record<string, unknown>> | undefined {
  if (!input.tools.length) return undefined
  const anchor = !Array.isArray(system)
    && (strategy === "default" || strategy === "tools-only")
  return input.tools.map((tool, index) => ({
    name: tool.name,
    description: tool.description,
    input_schema: JSON.parse(tool.parameters),
    ...(anchor && index === input.tools.length - 1
      ? { cache_control: { type: "ephemeral" } }
      : {}),
  }))
}

function countSlots(
  system: string | Array<Record<string, unknown>> | undefined,
  tools: Array<Record<string, unknown>> | undefined,
  messages: Array<Record<string, unknown>>,
): AnthropicRequestPlan["cacheSlots"] {
  return {
    system: Array.isArray(system) && system.some(block => block.cache_control),
    tools: !!tools?.some(tool => tool.cache_control),
    messages: messages.some(message =>
      Array.isArray(message.content)
      && (message.content as Array<Record<string, unknown>>).some(block => block.cache_control)),
  }
}

function countCacheBreakpoints(value: unknown): number {
  if (Array.isArray(value)) return value.reduce((total, item) => total + countCacheBreakpoints(item), 0)
  if (!value || typeof value !== "object") return 0
  const record = value as Record<string, unknown>
  return (record.cache_control ? 1 : 0)
    + Object.entries(record)
      .filter(([key]) => key !== "cache_control")
      .reduce((total, [, item]) => total + countCacheBreakpoints(item), 0)
}

function assertCacheBudget(params: Record<string, unknown>): void {
  const count = countCacheBreakpoints({
    system: params.system,
    tools: params.tools,
    messages: params.messages,
  })
  if (count > 4) {
    throw new Error(`Anthropic cache_control budget exceeded: ${count} > 4`)
  }
}

function numeric(raw: Record<string, unknown>, field: string): number | undefined {
  const value = raw[field]
  if (value === undefined || value === null) return undefined
  if (typeof value !== "number" || !Number.isSafeInteger(value) || value < 0) {
    throw new ProtocolResponseError("anthropic-messages", `usage.${field} must be a non-negative safe integer`)
  }
  return value
}

export class AnthropicMessagesAdapter implements ProtocolAdapter<
  AnthropicRequestPlan,
  Record<string, any>,
  AnthropicStreamChunk,
  AnthropicStreamState,
  undefined
> {
  readonly protocol = "anthropic-messages" as const
  readonly protocolCapabilities = ANTHROPIC_PROTOCOL_CAPABILITIES

  buildRequest(input: CanonicalAdapterInput): AnthropicRequestPlan {
    const strategy = cacheStrategy(input.extensions)
    const system = systemBlocks(input.context, strategy)
    const messages = input.context.turns
      .map(canonicalMessage)
      .filter((message): message is Record<string, unknown> => message !== undefined)
    applyMessageCacheControl(messages, input.context.frozenPrefixLen, strategy)
    if (input.context.stateTurn) {
      const stateMessage = canonicalMessage(input.context.stateTurn)
      if (stateMessage) messages.push(stateMessage)
    }
    if (!messages.length) messages.push({ role: "user", content: "Proceed." })
    const tools = requestTools(input, system, strategy)
    const wire = extensionsForWire(input.extensions)
    const betas = Array.isArray(input.extensions.betas) && input.extensions.betas.length
      ? input.extensions.betas
      : undefined
    const params = {
      ...wire,
      model: input.resolved.identity.modelId,
      max_tokens: typeof input.extensions.max_tokens === "number"
        ? input.extensions.max_tokens
        : 8096,
      ...(system ? { system } : {}),
      messages,
      ...(tools ? { tools } : {}),
      ...(betas ? { betas } : {}),
    }
    assertCacheBudget(params)
    return {
      transport: betas ? "beta" : "stable",
      params,
      cacheSlots: countSlots(system, tools, messages),
    }
  }

  decodeComplete(raw: Record<string, any>, decodeInput: AdapterDecodeInput): {
    message: ModelMessage
    replay?: ProviderReplay
  } {
    let content = ""
    const toolCalls: ToolCall[] = []
    for (const block of raw.content ?? []) {
      if (block.type === "text") content += block.text
      else if (block.type === "tool_use") {
        const call = normalizeToolCall(block.id, block.name, block.input)
        if (call) toolCalls.push(call)
      }
    }
    if (hasTextualToolCall(content, decodeInput.input)) throw textualToolCallError()
    const blocks = raw.content as Array<Record<string, unknown>> | undefined
    return {
      message: {
        role: "assistant",
        content,
        toolCalls,
      },
      ...(blocks?.length ? { replay: { protocol: "anthropic-messages" as const, native_blocks: blocks } } : {}),
    }
  }

  createStreamState(input: AdapterStreamInput): AnthropicStreamState {
    const plan = this.buildRequest(input.input)
    return {
      input: input.input,
      cacheSlots: plan.cacheSlots,
      toolBlocks: {},
      nativeBlocks: {},
      finalToolCalls: [],
      textualToolCallGuard: {
        policy: textualToolCallPolicy(input.input),
        mode: "passthrough",
        tail: "",
        capturedLength: 0,
        sawNativeToolCall: false,
      },
      finalText: "",
      uncachedInput: 0,
      cacheReadTokens: 0,
      cacheCreationTokens: 0,
      cacheTelemetryMeasured: false,
      outputTokens: 0,
    }
  }

  pushStreamChunk(chunk: AnthropicStreamChunk, state: AnthropicStreamState): AdapterOutput {
    const events: StreamEvent[] = []
    if (chunk.type === "message_start" || chunk.type === "message_delta") {
      const raw = chunk.usage ?? chunk.message?.usage
      if (raw) {
        if (
          Object.prototype.hasOwnProperty.call(raw, "cache_read_input_tokens")
          || Object.prototype.hasOwnProperty.call(raw, "cache_creation_input_tokens")
        ) {
          state.cacheTelemetryMeasured = true
        }
        state.uncachedInput = Math.max(state.uncachedInput, numeric(raw, "input_tokens") ?? 0)
        state.cacheReadTokens = Math.max(state.cacheReadTokens, numeric(raw, "cache_read_input_tokens") ?? 0)
        state.cacheCreationTokens = Math.max(state.cacheCreationTokens, numeric(raw, "cache_creation_input_tokens") ?? 0)
        state.outputTokens = Math.max(state.outputTokens, numeric(raw, "output_tokens") ?? 0)
        const inputTokens = state.uncachedInput + state.cacheReadTokens + state.cacheCreationTokens
        const cacheTelemetryStatus = state.cacheTelemetryMeasured ? "measured" : "unavailable"
        const providerUsage: ProviderUsage = {
          inputTokens,
          outputTokens: state.outputTokens,
          ...(state.cacheReadTokens ? { cacheReadInputTokens: state.cacheReadTokens } : {}),
          ...(state.cacheCreationTokens ? { cacheCreationInputTokens: state.cacheCreationTokens } : {}),
          cacheTelemetryStatus,
          ...(state.cacheTelemetryMeasured ? { cacheTelemetrySource: "anthropic_usage" } : {}),
        }
        const rawStopReason = chunk.delta?.stop_reason as string | undefined
        const stopReason = this.normalizeStopReason(rawStopReason)
        events.push({
          type: "usage",
          totalTokens: inputTokens + state.outputTokens,
          inputTokens,
          outputTokens: state.outputTokens,
          cacheReadInputTokens: state.cacheReadTokens,
          cacheCreationInputTokens: state.cacheCreationTokens,
          cacheTelemetryStatus,
          ...(state.cacheTelemetryMeasured ? { cacheTelemetrySource: "anthropic_usage" } : {}),
          ...(stopReason ? { stopReason } : {}),
          ...(rawStopReason ? { rawStopReason } : {}),
          providerUsage,
        } as UsageEvent)
      }
    } else if (chunk.type === "content_block_start") {
      state.nativeBlocks[chunk.index] = { ...chunk.content_block }
      if (chunk.content_block.type === "text") {
        const initialText = String(chunk.content_block.text ?? "")
        const visibleText = guardTextDelta(initialText, state.textualToolCallGuard)
        state.finalText += visibleText
        if (visibleText) events.push({ type: "text_delta", delta: visibleText } as StreamEvent)
      } else if (chunk.content_block.type === "tool_use") {
        state.textualToolCallGuard.sawNativeToolCall = true
        state.toolBlocks[chunk.index] = {
          id: chunk.content_block.id,
          name: chunk.content_block.name,
          argsBuffer: "",
        }
      }
    } else if (chunk.type === "content_block_delta") {
      const delta = chunk.delta
      if (delta.type === "text_delta") {
        const visibleText = guardTextDelta(delta.text, state.textualToolCallGuard)
        state.finalText += visibleText
        state.nativeBlocks[chunk.index] = {
          ...state.nativeBlocks[chunk.index],
          text: String(state.nativeBlocks[chunk.index]?.text ?? "") + delta.text,
        }
        if (visibleText) events.push({ type: "text_delta", delta: visibleText } as StreamEvent)
      } else if (delta.type === "thinking_delta") {
        state.nativeBlocks[chunk.index] = {
          ...state.nativeBlocks[chunk.index],
          thinking: String(state.nativeBlocks[chunk.index]?.thinking ?? "") + delta.thinking,
        }
        events.push({ type: "thinking_delta", delta: delta.thinking } as ThinkingDelta)
      } else if (delta.type === "signature_delta") {
        state.nativeBlocks[chunk.index] = {
          ...state.nativeBlocks[chunk.index],
          signature: String(state.nativeBlocks[chunk.index]?.signature ?? "") + delta.signature,
        }
      } else if (delta.type === "input_json_delta" && state.toolBlocks[chunk.index]) {
        state.toolBlocks[chunk.index].argsBuffer += delta.partial_json
      }
    } else if (chunk.type === "content_block_stop" && state.toolBlocks[chunk.index]) {
      const tool = state.toolBlocks[chunk.index]
      delete state.toolBlocks[chunk.index]
      let args: Record<string, unknown> = {}
      try { args = JSON.parse(tool.argsBuffer || "{}") as Record<string, unknown> } catch { args = {} }
      state.nativeBlocks[chunk.index] = { ...state.nativeBlocks[chunk.index], input: args }
      state.finalToolCalls.push({
        id: tool.id,
        name: tool.name,
        arguments: JSON.stringify(args),
      })
      events.push({
        type: "tool_call",
        id: tool.id,
        name: tool.name,
        arguments: args,
      } as ToolCallEvent)
    }
    return { events }
  }

  pushStreamChunkAtBoundary(request: AnthropicStreamChunkRequest): AdapterOutput {
    return this.pushStreamChunk(request.chunk, request.state)
  }

  finishStream(state: AnthropicStreamState): AdapterOutput {
    if (state.textualToolCallGuard.mode === "candidate") throw textualToolCallError()
    const events: StreamEvent[] = []
    if (state.textualToolCallGuard.tail) {
      const tail = state.textualToolCallGuard.tail
      state.textualToolCallGuard.tail = ""
      state.finalText += tail
      events.push({ type: "text_delta", delta: tail } as StreamEvent)
    }
    const blocks = Object.keys(state.nativeBlocks)
      .map(Number)
      .sort((left, right) => left - right)
      .map(index => state.nativeBlocks[index])
    return {
      events,
      ...(blocks.length ? { replay: { protocol: "anthropic-messages" as const, native_blocks: blocks } } : {}),
    }
  }

  finishStreamAtBoundary(request: AnthropicStreamFinishRequest): AdapterOutput {
    return this.finishStream(request.state)
  }

  normalizeUsage(raw: unknown): ProviderUsage | undefined {
    if (raw === undefined || raw === null) return undefined
    if (typeof raw !== "object" || Array.isArray(raw)) {
      throw new ProtocolResponseError("anthropic-messages", "usage must be an object")
    }
    const record = raw as Record<string, unknown>
    const uncached = numeric(record, "input_tokens")
    const cacheRead = numeric(record, "cache_read_input_tokens")
    const cacheCreation = numeric(record, "cache_creation_input_tokens")
    const output = numeric(record, "output_tokens")
    if (
      uncached === undefined && cacheRead === undefined
      && cacheCreation === undefined && output === undefined
    ) return undefined
    return {
      inputTokens: (uncached ?? 0) + (cacheRead ?? 0) + (cacheCreation ?? 0),
      outputTokens: output ?? 0,
      ...(cacheRead ? { cacheReadInputTokens: cacheRead } : {}),
      ...(cacheCreation ? { cacheCreationInputTokens: cacheCreation } : {}),
      ...(
        Object.prototype.hasOwnProperty.call(record, "cache_read_input_tokens")
        || Object.prototype.hasOwnProperty.call(record, "cache_creation_input_tokens")
          ? { cacheTelemetryStatus: "measured" as const, cacheTelemetrySource: "anthropic_usage" as const }
          : { cacheTelemetryStatus: "unavailable" as const }
      ),
    }
  }

  normalizeStopReason(raw: string | undefined): CanonicalStopReason | undefined {
    if (raw === undefined) return undefined
    switch (raw) {
      case "end_turn": return "end_turn"
      case "tool_use": return "tool_use"
      case "max_tokens": return "max_tokens"
      case "stop_sequence": return "stop_sequence"
      default: return "other"
    }
  }
}
